pauthabielf64/pauthabielf64.rst

```diff
@@ -542,243 +586,191 @@ by the run-time environment. There are a number of possible encoding
-+-----------+------------------------+------------+-----------------------------------------------------+
-
-This is the equivalent of the arm64e ARM64_RELOC_AUTHENTICATED relocation.
-
-
-GOT generating relocations
---------------------------
-
-* If the GOT is unsigned then no additional relocations are
-  required. The static linker creates unsigned entries as normal.
-
-* If the GOT is signed with an implicit schema then no new relocation
-  directives are required, but the existing GOT generating relocations
-  must encode the implicit signing schema in the GOT.
-
-  * This implies that there must be some way of the static linker
-    knowing to interpret the existing GOT relocations in a new
-    way. For example a command line option or metadata in the
-    relocatable objects.
-
-* If the GOT is signed with the source schema then the following GOT
-  generating relocations are used with the metadata encoded in the
-  relocation addend.
-
-Implicit signing schema
-^^^^^^^^^^^^^^^^^^^^^^^
-
-The GOT entries use the A key, a 0 discriminator with address
-diversity from the address of the GOT entry. No new relocation
-directives are needed. The static linker must encode the signing
-schema into the GOT slot.
-
-Code to access a GOT entry
-
-.. code-block:: asm
-
-  adrp x8, :got: symbol
-  add x8, x8, :got_lo12: symbol
-  ldr x0, [x8]
-  // Authenticate to get unsigned pointer
-  autia x0, x8
-  // Code should re-sign with source schema`
-
-Explicit signing schema
-^^^^^^^^^^^^^^^^^^^^^^^
++-----------+-------------------------+------------+-----------------------------------------------------+
+|ELF 64 Code| Name                    | Operation  | Comment                                             |
+|           |                         |            |                                                     |
+|           |                         |            |                                                     |
++-----------+-------------------------+------------+-----------------------------------------------------+
+| 0xE100    |R\_AARCH64\_AUTH\_ABS64  | PAUTH(S+A) | Signing schema encoded in the contents of the place |
++-----------+-------------------------+------------+-----------------------------------------------------+
 
-If the linker is to sign GOT entries with the source schema the
-following GOT generating relocations must be used. As the linker
-generates the pointer in the GOT the contents of the place of the
-relocation is an instruction so the signing schema cannot be encoded
-in the contents of the place. The relocation addend is used instead.
+This is the equivalent of the arm64e ARM64_RELOC_AUTHENTICATED
+relocation. It can also be used as a dynamic relocation.
 
-``ADDEND(A)`` extracts the addend field from the bottom 32-bits of the
-64 bit RELA addend field (signextend64, A & 0xffffffff)
+AUTH variant GOT Generating Relocations
```

Comment:
I still think we don't need to specify these because it is possible for the compiler to create these GOT entries by itself and rely on the linker to deduplicate them.

Reply:
The existing GOT generating relocations could be used in this context but this does require the compiler and linker to agree on whether the GOT is signed, and what subset of the GOT is signed. In a draft I had 3 options:

- No signing of the GOT (default)
- Code pointers in GOT are signed
- Code and Data pointers are signed
We then need a mechanism to communicate the option from the compiler to the static linker and there isn't a good existing option, a prototype could use a command line option but ideally this would be automatic. 

The .note.gnu.property isn't ideal for this purpose as this was intended to record architectural properties and doesn't handle anything other than feature bits that are subject to And or Or operations. 

Something like AArch32 BuildAttributes could work but that would mean specifying them.

Using AUTH GOT generating relocations handles the communication without needing some kind of binary marking.
- Unsigned GOT = usual GOT generating relocations
- Code pointers = mix of AUTH and usual GOT generating relocations
- Code and Data pointers = AUTH GOT generating relocations

Hope I haven't misunderstood your comment?

I think we may be lucky and the initial adoption doesn't need any kind of signed GOT. Experiments could use a command line option.